### PR TITLE
Add support for "BKS" (BouncyCastle) format keystore files

### PR DESCRIPTION
**Pull request by kholia**
The PKCS #12 KDF will be useful for cracking PFX files as well.

### CODE REVIEW

**Comment by kholia** [comment]
I am running into the following problem,

```
$ ../run/john --format=bks --test-full=0
Testing: BKS [PKCS12 PBE SHA-1 32/64]... FAILED (should not set FMT_SPLIT_UNIFIES_CASE (#4))
```

@magnumripper what does this mean? :)

**Comment by magnumripper** [comment]
It's pretty clear in the output. You shouldn't set FMT_SPLIT_UNIFIES_CASE unless your split() function actually unifies case. This test was made just for you, my friend 😉

**Comment by magnumripper** [comment]
Hmmm but if this is the format you just committed, I'm very puzzled.

File: src/bks_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+		BENCHMARK_LENGTH,
+		0,
+		PLAINTEXT_LENGTH,
+		BINARY_SIZE,
+		BINARY_ALIGN,
+		SALT_SIZE,
+		SALT_ALIGN,
+		MIN_KEYS_PER_CRYPT,
+		MAX_KEYS_PER_CRYPT,
+		// FMT_NOT_EXACT is required only for BKS keystores
```
Comment (magnumripper) [inline src/bks_fmt_plug.c]: Does this format support different types and just one of them is FMT_NOT_EXACT? In that case you could opt to clear this flag and then set it on a per-salt basis instead.

**Comment by magnumripper** [comment]
I get it too. You need to run `-test-full=0` to trigger it.
```
$ ../run/john -test-full=0 -form:bks
Will run 8 OpenMP threads
Testing: bks [PKCS12 PBE SHA-1 32/64]... (8xOMP) FAILED (should not set FMT_SPLIT_UNIFIES_CASE (#4))
```

**Comment by kholia** [comment]
I haven't set `FMT_SPLIT_UNIFIES_CASE` in my format. This is a puzzling failure.

**Comment by kholia** [comment]
I am also unable to use the inbuilt fuzzing facility,

```
$ ../run/john --format=bks --fuzz
Fuzzing: BKS [PKCS12 PBE SHA-1 32/64]... 
Created directory: fuzz_status
ciphertext='$bks$0$1$20$2036$...' is bigger than the LINE_BUFFER_SIZE=1024
```

**Comment by magnumripper** [comment]
This smells like ASan or Valgrind would find a problem that leads to this strange things.

**Comment by kholia** [comment]
ASan is enabled on my builds, and it is not being triggered at all.

**Comment by kholia** [comment]
This PR now seems to be ready for merge :-)

**Comment by magnumripper** [comment]
👍

**Comment by magnumripper** [comment]
For finding BKS collisions, user can run with `--keep-guessing`. Only formats that may produce false (unusable) positives should set FMT_NOT_EXACT. So all is fine!

**Pull request merged**